Objects/typeobject.c

```diff
@@ -2328,10 +2333,17 @@ extra_ivars(PyTypeObject *type, PyTypeObject *base)
         return t_size != b_size ||
             type->tp_itemsize != base->tp_itemsize;
     }
-    if (type->tp_weaklistoffset && base->tp_weaklistoffset == 0 &&
-        type->tp_weaklistoffset + sizeof(PyObject *) == t_size &&
-        type->tp_flags & Py_TPFLAGS_HEAPTYPE)
+    /* Check for __dict__ and __weakrefs__ slots in either order */
+    if (ADDED_FIELD_AT_OFFSET(weaklistoffset, t_size)) {
+        t_size -= sizeof(PyObject *);
+    }
+    if ((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0 &&
+        ADDED_FIELD_AT_OFFSET(dictoffset, t_size)) {
         t_size -= sizeof(PyObject *);
+    }
+    if (ADDED_FIELD_AT_OFFSET(weaklistoffset, t_size)) {
+        t_size -= sizeof(PyObject *);
+    }
```

Comment:
Why are we doing `weaklistoffset` twice?

Reply:
So we can handle `weakreflist` followed by `dictoffset` and `dictoffset` followed by  `weakreflist`.

Reply:
Ah, right.  The first one won't match the current `t_size` if `dictoffset` comes first.  Perhaps that would be more clear like this:

```c
    if ((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0 &&
        ADDED_FIELD_AT_OFFSET(dictoffset, t_size)) {
        t_size -= sizeof(PyObject *);

        /* Check __weakrefs__ again, in case it precedes __dict__ */
        if (ADDED_FIELD_AT_OFFSET(weaklistoffset, t_size)) {
            t_size -= sizeof(PyObject *);
        }
    }
```

Then again, it would be especially clear like this:

```c
    /* Check for __dict__ and __weakrefs__ slots in either order */
    if (ADDED_FIELD_AT_OFFSET(weaklistoffset, t_size)) {
        t_size -= sizeof(PyObject *);

        if ((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0 &&
            ADDED_FIELD_AT_OFFSET(dictoffset, t_size)) {
            t_size -= sizeof(PyObject *);
        }
    }
    else if ((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0 &&
        ADDED_FIELD_AT_OFFSET(dictoffset, t_size)) {
        t_size -= sizeof(PyObject *);

        if (ADDED_FIELD_AT_OFFSET(weaklistoffset, t_size)) {
            t_size -= sizeof(PyObject *);
        }
    }
```

Reply:
Regardless, what you have is fine.